transit_api_511.py

```diff
@@ -0,0 +1,213 @@
+"""
+All the functionality needed to interact with the 511.org API and parse the data is here.
+
+API reference:
+https://511.org/open-data/transit
+
+Current reference:
+https://511.org/sites/default/files/2022-11/511%20SF%20Bay%20Open%20Data%20Specification%20-%20Transit.pdf
+
+In order to obtain an access token to use for this API go to:
+https://511.org/open-data/token
+"""
+
+from json import loads
+
+# libs
+from adafruit_datetime import datetime
+
+BASE_URL = 'api.511.org'
+
+
+class JsonHandler:
+    """
+    Parses the data returned in JSON format from the 511.org API.
+    """
+
+    @staticmethod
+    def parse_data(data):
+        """
+        Takes the cleartext response data from an API call and formats it to a JSON map
+
+        :return: the data in JSON format
+        """
+
+        return loads(data)
+
+    @staticmethod
+    def predictions_for_route_codes(json, route_codes, direction):
+        """
+        Takes all the prediction JSON and collects only the relevant items.
+
+        :param json: the data to parse
+        :param route_codes: the routes of interest
+        :param direction: the direction of interest
+        :return: the predictions that are available if any
+        """
+
+        predictions = {}
+
+        # The time the predictions data was sent
+        # This is used because some predictions come back with a 'created' unix time stamp
+        # of 0, so they are inaccurate to use for temporal subtraction so a unified baseline
+        # of 'now' is used to get around this flaw in the API response
+        data_time = json['ServiceDelivery']['StopMonitoringDelivery']['ResponseTimestamp']
+        # CircuitPython only parses DateTime from iso formats without the 'T' and 'Z' in them.
+        data_time = data_time.replace('T', ' ').replace('Z', '')
+        now = datetime.fromisoformat(data_time)
+
+        # The list of upcoming visits to the stop in the predictions
+        visits = json['ServiceDelivery']['StopMonitoringDelivery']['MonitoredStopVisit']
+
+        # Iterate over all the upcoming visits and gather predictions
+        for visit in visits:
+            trip = visit['MonitoredVehicleJourney']
+            route_code = trip['LineRef']
+
+            # If the route code is of interest in the trip in the correct direction
+            if route_code in route_codes and trip['DirectionRef'] == direction:
```

Comment:
Consider lowercase or uppercase conversion

Reply:
I am not sure what you mean here

Reply:
I meant to handle case sensitivity